src/coprocessor/dag/batch/executors/stream_aggr_executor.rs

```diff
@@ -0,0 +1,575 @@
+// Copyright 2019 TiKV Project Authors. Licensed under Apache-2.0.
+
+use std::convert::TryFrom;
+use std::sync::Arc;
+
+use cop_datatype::{EvalType, FieldTypeAccessor};
+use tipb::executor::Aggregation;
+use tipb::expression::{Expr, FieldType};
+
+use crate::coprocessor::codec::batch::{LazyBatchColumn, LazyBatchColumnVec};
+use crate::coprocessor::codec::data_type::*;
+use crate::coprocessor::codec::mysql::time::Tz;
+use crate::coprocessor::dag::aggr_fn::*;
+use crate::coprocessor::dag::batch::executors::util::aggr_executor::*;
+use crate::coprocessor::dag::batch::interface::*;
+use crate::coprocessor::dag::expr::{EvalConfig, EvalContext};
+use crate::coprocessor::dag::rpn_expr::types::RpnStackNode;
+use crate::coprocessor::dag::rpn_expr::{RpnExpression, RpnExpressionBuilder};
+use crate::coprocessor::Result;
+
+pub struct BatchStreamAggregationExecutor<Src: BatchExecutor>(
+    AggregationExecutor<Src, BatchStreamAggregationImpl>,
+);
+
+impl<Src: BatchExecutor> BatchExecutor for BatchStreamAggregationExecutor<Src> {
+    #[inline]
+    fn schema(&self) -> &[FieldType] {
+        self.0.schema()
+    }
+
+    #[inline]
+    fn next_batch(&mut self, scan_rows: usize) -> BatchExecuteResult {
+        self.0.next_batch(scan_rows)
+    }
+
+    #[inline]
+    fn collect_statistics(&mut self, destination: &mut BatchExecuteStatistics) {
+        self.0.collect_statistics(destination)
+    }
+}
+
+impl<Src: BatchExecutor> BatchStreamAggregationExecutor<Src> {
+    #[cfg(test)]
+    pub fn new_for_test(
+        src: Src,
+        group_by_exps: Vec<RpnExpression>,
+        aggr_defs: Vec<Expr>,
+        aggr_def_parser: impl AggrDefinitionParser,
+    ) -> Self {
+        Self::new_impl(
+            Arc::new(EvalConfig::default()),
+            src,
+            group_by_exps,
+            aggr_defs,
+            aggr_def_parser,
+        )
+        .unwrap()
+    }
+}
+
+impl BatchStreamAggregationExecutor<Box<dyn BatchExecutor>> {
+    /// Checks whether this executor can be used.
+    #[inline]
+    pub fn check_supported(descriptor: &Aggregation) -> Result<()> {
+        let group_by_definitions = descriptor.get_group_by();
+        assert!(!group_by_definitions.is_empty());
+        for def in group_by_definitions {
+            RpnExpressionBuilder::check_expr_tree_supported(def)?;
+            if RpnExpressionBuilder::is_expr_eval_to_scalar(def)? {
+                return Err(box_err!("Group by expression cannot be a scalar"));
```

Comment:
Are you sure the group by item `"hello"` will not be pushed down in `SELECT c1 FROM t GROUP BY c1, "hello"?

Reply:
Nice catch. This case was not considered. SlowHashAggr also need to be fixed.

Reply:
So only when there is one group by column, it cannot be a constant. When there are multiple group by columns, constant is possible. Right?

Reply:
Looks like TiDB will not push down "hello" to TiKV:

```
mysql> explain SELECT c1 FROM t GROUP BY c1, "hello";
+-----------------------+----------+------+------------------------------------------------------------+
| id                    | count    | task | operator info                                              |
+-----------------------+----------+------+------------------------------------------------------------+
| HashAgg_9             | 8000.00  | root | group by:col_1, funcs:firstrow(col_0)                      |
| └─TableReader_10      | 8000.00  | root | data:HashAgg_5                                             |
|   └─HashAgg_5         | 8000.00  | cop  | group by:test.t.c1, funcs:firstrow(test.t.c1)              |
|     └─TableScan_8     | 10000.00 | cop  | table:t, range:[-inf,+inf], keep order:false, stats:pseudo |
+-----------------------+----------+------+------------------------------------------------------------+
4 rows in set (0.00 sec)
```

Reply:
So, it's ok to keep it an error?

Reply:
Looks right. @xuhuaiyu Could you help us confirm? We need to know whether or not TiDB will not push down the constant part in group by. Thanks!

Reply:
Seems TiDB will never push down a constant, See: https://github.com/pingcap/tidb/blob/9071ab9131bce65ebf08a879807d2b9c163536a4/planner/core/rule_column_pruning.go#L124